File: leetcode/python3/binary-tree-inorder-traversal.py
#
# @lc app=leetcode id=94 lang=python3
#
# [94] Binary Tree Inorder Traversal
#
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        # ans = []
        # def inorder_traversal(node):        
        #     if node:
        #         inorder_traversal(node.left)
        #         ans.append(node.val)
        #         inorder_traversal(node.right)
        # inorder_traversal(root)
        # return ans
        ans = []
        stack = []
        cur = root
        while cur or stack:
            if cur:
                stack.append(cur)
                cur = cur.left
                continue
            tmp = stack.pop()
            ans.append(tmp.val)
            cur = tmp.right
        return ans

File: leetcode/cpp/binary-tree-inorder-traversal.cpp
/*
 * @lc app=leetcode id=94 lang=cpp
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        stack<TreeNode*> s;
        TreeNode* p = root;
        vector<int> ans;
        while (p || !s.empty()) {
            while (p) {
                s.push(p);
                p = p->left;
            }

            if (!s.empty()) {
                p = s.top();
                ans.push_back(p->val);
                s.pop();
                p = p->right;
            }
        }
        return ans;
    }
};
// @lc code=end

